tool/tbot/config/configtemplate.go

```diff
@@ -0,0 +1,118 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package config
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/teleport/tool/tbot/destination"
+	"github.com/gravitational/teleport/tool/tbot/identity"
+	"github.com/gravitational/trace"
+	"gopkg.in/yaml.v3"
+)
+
+const TemplateSSHClientName = "ssh_client"
+
+// AllConfigTemplates lists all valid config templates, intended for help
+// messages
+var AllConfigTemplates = [...]string{TemplateSSHClientName}
+
+// FileDescription is a minimal spec needed to create an empty end-user-owned
+// file with bot-writable ACLs during `tbot init`.
+type FileDescription struct {
+	// Name is the name of the file or directory to create.
+	Name string
+
+	// IsDir designates whether this describes a subdirectory inside the
+	// destination.
+	IsDir bool
+
+	// ModeHint describes the intended permissions for this data, for
+	// Destination backends where permissions are relevant.
+	ModeHint destination.ModeHint
+}
+
+// Template defines functions for dynamically writing additional files to
+// a Destination.
+type Template interface {
+	// Describe generates a list of all files this ConfigTemplate will generate
+	// at runtime. Currently ConfigTemplates are required to know this
+	// statically as this must be callable without any auth clients (or any
+	// secrets) for use with `tbot init`. If an arbitrary number of files must
+	// be generated, they should be placed in a subdirectory.
+	Describe() []FileDescription
+
+	// Render writes the config template to the destination.
+	Render(authClient *auth.Client, currentIdentity *identity.Identity, destination *DestinationConfig) error
+}
+
+// TemplateConfig contains all possible config template variants. Exactly one
+// variant must be set to be considered valid.
+type TemplateConfig struct {
+	SSHClient *TemplateSSHClient `yaml:"ssh_client,omitempty"`
+}
+
+func (c *TemplateConfig) UnmarshalYAML(node *yaml.Node) error {
+	var simpleTemplate string
+	if err := node.Decode(&simpleTemplate); err == nil {
+		switch simpleTemplate {
+		case TemplateSSHClientName:
+			c.SSHClient = &TemplateSSHClient{}
+			fmt.Println("no params, using defaults")
+		default:
+			return trace.BadParameter(
+				"invalid config template '%s' on line %d, expected one of: %s",
+				simpleTemplate, node.Line, strings.Join(AllConfigTemplates[:], ", "),
+			)
+		}
+		return nil
+	}
+
+	type rawTemplate TemplateConfig
+	if err := node.Decode((*rawTemplate)(c)); err != nil {
+		return err
+	}
+
+	return nil
```

Comment:
I don't understand the typecast, could probably use a comment.

Also the `if` is doing nothing and fails lint, you can just `return trace.Wrap(err)`

Reply:
Ah, I just didn't copy and paste the comment everywhere. It's derived from https://abhinavg.net/posts/flexible-yaml/#make-it-flexible to allow optional parameters for templates